prog/dftb+/lib_dftb/dispdftd4.F90

```diff
@@ -136,55 +183,113 @@ subroutine DispDftD4_init(this, inp, nAtom, speciesNames, latVecs)
     allocate(this%energies(nAtom))
     allocate(this%gradients(3, nAtom))
 
-    allocate(this%calculator)
-    call initializeCalculator(this%calculator, inp)
+    call init_(this%calc, inp)
+    call init_(this%ref, inp%izp, inp%chargeScale, inp%chargeSteepness)
+
+    ! Calculation must be self-consistent if no charge model is available
+    if (inp%tSelfConsistent) then
+      allocate(this%sc)
+      call TScD4_init(this%sc, this%ref, nAtom)
+    end if
 
   end subroutine DispDftD4_init
 
 
-  !> Notifies the objects about changed coordinates.
+  !> Initialize dispersion data cache for self-consistent evaluation
+  subroutine TScD4_init(this, ref, nAtom)
+
+    !> Instance of the self-consistent dispersion data
+    type(TScD4), intent(out) :: this
+
+    !> DFT-D4 reference systems
+    type(TDftD4Ref), intent(in) :: ref
+
+    !> Number of atoms
+    integer, intent(in) :: nAtom
+
+    integer :: nRef
+
+    nRef = maxval(ref%nRef)
+    this%nRef = nRef
+
+    allocate(this%charges(nAtom))
+    allocate(this%shift(nAtom))
+    allocate(this%energies(nAtom))
+    allocate(this%dispmat(nRef, nAtom, nRef, nAtom))
+    allocate(this%scratch(nRef, nAtom))
+    allocate(this%zetaVec(nRef, nAtom))
+    allocate(this%zetadq(nRef, nAtom))
+
+  end subroutine TScD4_init
+
+
+  !> Write information about DFT-D4 dispersion model
+  subroutine writeDftD4Info(unit, this)
+    !> Formatted unit for output
+    integer, intent(in) :: unit
+
+    !> Instance of the type to describe
+    class(TDispDftD4), intent(in) :: this
+
+    if (allocated(this%sc)) then
+      write(unit, "(A)") "Using self-consistent DFT-D4 dispersion corrections"
+    else
+      write(unit, "(A)") "Using DFT-D4 dispersion corrections"
+    end if
+  end subroutine writeDftD4Info
```

Comment:
What about making this a type bound procedure, returning `.true.` or `.false.` and letting the caller decide about the best way to render this information. Alternatively, returning an allocatable character with `"self-consistent"` or `"non-self-consistent"` would make also sense to me, but still, with rendering on the caller side. (So, that in a far future we can write a good unit test for this functionality as well :wink:)

Reply:
The solvation model uses a similar strategy, the containers know how to display their information correctly, the main program only knows that the containers can display this information. This strategy is necessary in case we have to access module private information for display. Returning a single string is not really a solution if the container becomes more complex.

Reply:
OK, I see. Than just leave this for the moment as it is.